File: sdks/python/src/gram_ai/function_calling.py
from dataclasses import dataclass
import functools
import json
from typing import Callable, Dict, Optional, Awaitable, Union

import httpx

from gram_ai import VERSION, GramAPI
from gram_ai.environments import get_server_url_by_key
from gram_ai.models.getinstanceresult import GetInstanceResult
from gram_ai.utils.retries import (
    BackoffStrategy,
    Retries,
    RetryConfig,
    retry_async,
    retry,
)


@dataclass
class GramFunctionCallingCall:
    tool_id: str
    project: str
    toolset: str
    environment: Optional[str] = None


@dataclass
class GramFunctionCallingTool:
    name: str
    description: str
    parameters: Dict
    execute: Callable[[str], str]
    aexecute: Callable[[str], Awaitable[str]]


class GramFunctionCalling:
    api_key: str
    server_url: str
    _cache: dict[tuple[str, str, Union[str, None]], list[GramFunctionCallingTool]] = {}

    def __init__(
        self,
        *,
        api_key: str,
    ):
        self.api_key = api_key
        self.server_url = get_server_url_by_key(api_key)
        self.client = GramAPI(server_url=self.server_url)

    async def _do_http_async(self, req: httpx.Request) -> httpx.Response:
        async with httpx.AsyncClient() as client:
            return await client.send(req)

    def _do_http_sync(self, req: httpx.Request) -> httpx.Response:
        with httpx.Client() as client:
            return client.send(req)

    def _fetch_tools(
        self,
        project: str,
        toolset: str,
        environment: Optional[str] = None,
    ) -> GetInstanceResult:
        return self.client.instances.get_by_slug(
            security={
                "option2": {
                    "apikey_header_gram_key": self.api_key,
                    "project_slug_header_gram_project": project,
                }
            },
            toolset_slug=toolset,
            environment_slug=environment,
        )

    def tools(
        self,
        project: str,
        toolset: str,
        environment: Optional[str] = None,
    ) -> list[GramFunctionCallingTool]:
        key = (project, toolset, environment)
        if key in self._cache:
            return self._cache[key]

        instance = self._fetch_tools(project, toolset, environment)

        tools: list[GramFunctionCallingTool] = []

        for tool in instance.tools:
            schema = json.loads(tool.schema_) if tool.schema_ else {}

            tool_call = GramFunctionCallingCall(tool.id, project, toolset, environment)

            func = self._create_sync_tool_function(tool_call)
            async_func = self._create_tool_function(tool_call)

            tools.append(
                GramFunctionCallingTool(
                    name=tool.name,
                    description=tool.description,
                    parameters=schema,
                    execute=func,
                    aexecute=async_func,
                )
            )

        self._cache[key] = tools
        return tools

    def _prepare_request(self, tool_call: GramFunctionCallingCall, inp: str):
        url = f"{self.server_url}/rpc/instances.invoke/tool"
        params = {"tool_id": tool_call.tool_id}
        if tool_call.environment:
            params["environment_slug"] = tool_call.environment

        json_input = json.loads(inp)

        return (
            url,
            params,
            {
                "gram-key": self.api_key,
                "gram-project": tool_call.project,
                "user-agent": f"gram-ai/openai-agents python {VERSION}",
                "content-type": "application/json",
            },
            json_input,
        )

    def _create_tool_function(self, tool_call: GramFunctionCallingCall):
        async def wrapper(inp: str):
            url, params, headers, data = self._prepare_request(tool_call, inp)
            req = httpx.Request(
                method="POST",
                url=url,
                params=params,
                headers=headers,
                json=data,
            )
            response = await retry_async(
                functools.partial(self._do_http_async, req), _retry_policy
            )
            response.raise_for_status()
            return response.text

        return wrapper

    def _create_sync_tool_function(self, tool_call: GramFunctionCallingCall):
        def wrapper(inp: str):
            url, params, headers, data = self._prepare_request(tool_call, inp)
            req = httpx.Request(
                method="POST",
                url=url,
                params=params,
                headers=headers,
                json=data,
            )
            response = retry(functools.partial(self._do_http_sync, req), _retry_policy)
            response.raise_for_status()
            return response.text

        return wrapper


_retry_policy = Retries(
    config=RetryConfig(
        strategy="backoff",
        backoff=BackoffStrategy(
            initial_interval=500,
            max_interval=60000,
            exponent=1.5,
            max_elapsed_time=3600000,
        ),
        retry_connection_errors=True,
    ),
    status_codes=["429", "5XX"],
)


__all__ = ["GramFunctionCalling"]

File: sdks/typescript/src/functioncalling.ts
import { GramAPICore } from "./core.js";
import { getServerUrlByKey } from "./environments.js";
import { instancesGetBySlug } from "./funcs/instancesGetBySlug.js";
import { isBrowserLike } from "./lib/browsers.js";
import { SDK_METADATA } from "./lib/config.js";
import { GetInstanceResult } from "./models/components/getinstanceresult.js";
import { unwrapAsync } from "./types/fp.js";

export type GramInstanceRequest = {
  project: string;
  toolset: string;
  environment?: string;
};

export type FunctionCallingTool = {
  name: string;
  description: string;
  parameters: object;
  execute: (input: string) => Promise<string>;
};

export class FunctionCallingAdapter {
  readonly #apiKey: string;
  readonly #serverURL: string;
  readonly #cache: Map<string, FunctionCallingTool[]> = new Map();
  readonly #core: GramAPICore;

  constructor(apiKey: string) {
    this.#apiKey = apiKey;
    this.#serverURL = getServerUrlByKey(apiKey);
    this.#core = new GramAPICore({
      serverURL: this.#serverURL,
    });
  }

  async #fetchInstance(
    project: string,
    toolset: string,
    environment?: string | undefined
  ): Promise<GetInstanceResult> {
    return unwrapAsync(
      instancesGetBySlug(
        this.#core,
        {
          option2: {
            apikeyHeaderGramKey: this.#apiKey,
            projectSlugHeaderGramProject: project,
          },
        },
        {
          toolsetSlug: toolset,
          environmentSlug: environment,
        }
      )
    );
  }

  async tools({
    project,
    toolset,
    environment,
  }: GramInstanceRequest): Promise<FunctionCallingTool[]> {
    const key = `${project}:${toolset}:${environment || ""}`;

    if (this.#cache.has(key)) {
      return this.#cache.get(key)!;
    }

    const client = this.#core;
    const apiKey = this.#apiKey;
    const instance = await this.#fetchInstance(project, toolset, environment);

    const tools: FunctionCallingTool[] = [];

    for (const toolData of instance.tools) {
      const schema = toolData.schema ? JSON.parse(toolData.schema) : {};
      const execute = async (input: string) => {
        const security: Record<string, string> = {
          "gram-key": apiKey,
          "gram-project": project,
        };
        const headers: Record<string, string> = { ...security };

        if (isBrowserLike) {
          headers[
            "user-agent"
          ] = `gram-ai/vercel typescript ${SDK_METADATA.sdkVersion}`;
        }

        const retryConfig = {
          strategy: "backoff",
          retryConnectionErrors: true,
        } as const;

        const url = new URL(
          `${this.#serverURL}/rpc/instances.invoke/tool`
        );
        url.searchParams.set("tool_id", toolData.id);
        if (environment) {
          url.searchParams.set("environment_slug", environment);
        }
        const request = new Request(url, {
          method: "POST",
          headers,
          body: input,
        });

        const result = await client._do(request, {
          context: {
            baseURL: this.#serverURL,
            operationID: "invokeTool",
            oAuth2Scopes: null,
            retryConfig,
            resolvedSecurity: {
              headers: security,
              basic: {},
              queryParams: {},
              cookies: {},
              oauth2: { type: "none" },
            },
          },
          errorCodes: ["4XX", "5XX"],
          retryConfig,
          retryCodes: ["5XX"],
        });
        if (!result.ok) {
          throw new Error(`Tool call failed: ${toolData.name}`, {
            cause: result.error,
          });
        }

        const response = result.value;
        if (!response.ok) {
          const body = await response.text().catch(() => "");

          throw new Error(
            `Tool call failed: ${toolData.name}: ${response.statusText}: ${body}`
          );
        }

        const json = await response.json();
        return JSON.stringify(json);
      };

      const functionCallingTool = {
        name: toolData.name,
        description: toolData.description,
        parameters: schema,
        execute,
      };

      tools.push(functionCallingTool);
    }

    this.#cache.set(key, tools);

    return tools;
  }
}
